panic in Writer run loop

**Describe the bug**

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x9e7650]

goroutine 85 [running]:
github.com/segmentio/kafka-go.(*Writer).run(0xc0000f8b40)
        /data/orange-ci/workspace/vendor/github.com/segmentio/kafka-go/writer.go:488 +0x560
created by github.com/segmentio/kafka-go.NewWriter
        /data/orange-ci/workspace/vendor/github.com/segmentio/kafka-go/writer.go:272 +0x2bd
```

**Kafka Version**
0.10

kafka-go version: v0.3.4
```go
if len(partitions) != 0 {
    selectedPartition := w.config.Balancer.Balance(wm.msg, partitions...)
    writers[selectedPartition].messages() <- wm
} 
```
it seams that the selectedPartition is not in writers, but i dont know why.

--- a/balancer.go
+++ b/balancer.go
@@ -67,11 +67,8 @@ type leastBytesCounter struct {
 
 // Balance satisfies the Balancer interface.
 func (lb *LeastBytes) Balance(msg Message, partitions ...int) int {
-	for _, p := range partitions {
-		if c := lb.counterOf(p); c == nil {
-			lb.counters = lb.makeCounters(partitions...)
-			break
-		}
+	if lb.partitionChanged(partitions...) {
+		lb.counters = lb.makeCounters(partitions...)
 	}
 
 	minBytes := lb.counters[0].bytes
@@ -87,6 +84,18 @@ func (lb *LeastBytes) Balance(msg Message, partitions ...int) int {
 	c := &lb.counters[minIndex]
 	c.bytes += uint64(len(msg.Key)) + uint64(len(msg.Value))
 	return c.partition
+}
+
+func (lb *LeastBytes) partitionChanged(partitions ...int) bool {
+	if len(partitions) != len(lb.counters) {
+		return true
+	}
+	for _, p := range partitions {
+		if c := lb.counterOf(p); c == nil {
+			return true
+		}
+	}
+	return false
 }
 
 func (lb *LeastBytes) counterOf(partition int) *leastBytesCounter {

--- a/balancer.go
+++ b/balancer.go
@@ -67,7 +67,8 @@ type leastBytesCounter struct {
 
 // Balance satisfies the Balancer interface.
 func (lb *LeastBytes) Balance(msg Message, partitions ...int) int {
-	if lb.partitionChanged(partitions...) {
+	// partitions change
+	if len(partitions) != len(lb.counters) {
 		lb.counters = lb.makeCounters(partitions...)
 	}
 
@@ -84,28 +85,6 @@ func (lb *LeastBytes) Balance(msg Message, partitions ...int) int {
 	c := &lb.counters[minIndex]
 	c.bytes += uint64(len(msg.Key)) + uint64(len(msg.Value))
 	return c.partition
-}
-
-func (lb *LeastBytes) partitionChanged(partitions ...int) bool {
-	if len(partitions) != len(lb.counters) {
-		return true
-	}
-	for _, p := range partitions {
-		if c := lb.counterOf(p); c == nil {
-			return true
-		}
-	}
-	return false
-}
-
-func (lb *LeastBytes) counterOf(partition int) *leastBytesCounter {
-	i := sort.Search(len(lb.counters), func(i int) bool {
-		return lb.counters[i].partition >= partition
-	})
-	if i == len(lb.counters) || lb.counters[i].partition != partition {
-		return nil
-	}
-	return &lb.counters[i]
 }
 
 func (lb *LeastBytes) makeCounters(partitions ...int) (counters []leastBytesCounter) {

--- a/balancer_test.go
+++ b/balancer_test.go
@@ -279,47 +279,81 @@ func TestMurmur2Balancer(t *testing.T) {
 }
 
 func TestLeastBytes(t *testing.T) {
-	testCases := []struct {
-		Label      string
-		Key        []byte
+	testNormalCases := map[string]struct {
+		Keys       [][]byte
 		Partitions []int
 		Partition  int
 	}{
-		{
-			Label:      "init",
-			Key:        []byte("key"),
-			Partitions: []int{1, 2, 3},
-			Partition:  1,
+		"one msg": {
+			Keys: [][]byte{
+				[]byte("key"),
+			},
+			Partitions: []int{0, 1, 2},
+			Partition:  0,
 		},
-		{
-			Label:      "normal",
-			Key:        []byte("key"),
-			Partitions: []int{1, 2, 3},
-			Partition:  2,
-		},
-		{
-			Label:      "partition-scale-down",
-			Key:        []byte("key"),
-			Partitions: []int{1, 2},
-			Partition:  1,
-		},
-		{
-			Label:      "partition-scale-up",
-			Key:        []byte("key"),
-			Partitions: []int{1, 2, 3},
+		"two msg": {
+			Keys: [][]byte{
+				[]byte("key"),
+				[]byte("key"),
+			},
+			Partitions: []int{0, 1, 2},
 			Partition:  1,
 		},
 	}
 
-	t.Run("default", func(t *testing.T) {
-		lb := &LeastBytes{}
-		for _, test := range testCases {
-			t.Run(test.Label, func(t *testing.T) {
-				msg := Message{Key: test.Key}
-				partition := lb.Balance(msg, test.Partitions...)
+	testPartitionChange := map[string]struct {
+		Key            []byte
+		PartitionsHist [][]int
+		Partition      int
+	}{
+		"reduce": {
+			Key: []byte("key"),
+			PartitionsHist: [][]int{
+				{0, 1, 2},
+				{0, 1},
+			},
+			Partition: 0,
+		},
+		"expand": {
+			Key: []byte("key"),
+			PartitionsHist: [][]int{
+				{0, 1},
+				{0, 1, 2},
+			},
+			Partition: 0,
+		},
+	}
+
+	t.Run("normal", func(t *testing.T) {
+		for label, test := range testNormalCases {
+			t.Run(label, func(t *testing.T) {
+				lb := &LeastBytes{}
+				var partition int
+				for _, key := range test.Keys {
+					msg := Message{Key: key}
+					partition = lb.Balance(msg, test.Partitions...)
+
+				}
 				if partition != test.Partition {
 					t.Errorf("expected %v; got %v", test.Partition, partition)
 				}
+			})
+		}
+	})
+
+	t.Run("partitionChange", func(t *testing.T) {
+		for label, test := range testPartitionChange {
+			t.Run(label, func(t *testing.T) {
+				lb := &LeastBytes{}
+				msg := Message{Key: test.Key}
+				var partition int
+				for _, partitions := range test.PartitionsHist {
+					partition = lb.Balance(msg, partitions...)
+				}
+				if partition != test.Partition {
+					t.Errorf("expected %v; got %v", test.Partition, partition)
+				}
+
 			})
 		}
 	})

--- a/balancer_test.go
+++ b/balancer_test.go
@@ -277,3 +277,50 @@ func TestMurmur2Balancer(t *testing.T) {
 		}
 	})
 }
+
+func TestLeastBytes(t *testing.T) {
+	testCases := []struct {
+		Label      string
+		Key        []byte
+		Partitions []int
+		Partition  int
+	}{
+		{
+			Label:      "init",
+			Key:        []byte("key"),
+			Partitions: []int{1, 2, 3},
+			Partition:  1,
+		},
+		{
+			Label:      "normal",
+			Key:        []byte("key"),
+			Partitions: []int{1, 2, 3},
+			Partition:  2,
+		},
+		{
+			Label:      "partition-scale-down",
+			Key:        []byte("key"),
+			Partitions: []int{1, 2},
+			Partition:  1,
+		},
+		{
+			Label:      "partition-scale-up",
+			Key:        []byte("key"),
+			Partitions: []int{1, 2, 3},
+			Partition:  1,
+		},
+	}
+
+	t.Run("default", func(t *testing.T) {
+		lb := &LeastBytes{}
+		for _, test := range testCases {
+			t.Run(test.Label, func(t *testing.T) {
+				msg := Message{Key: test.Key}
+				partition := lb.Balance(msg, test.Partitions...)
+				if partition != test.Partition {
+					t.Errorf("expected %v; got %v", test.Partition, partition)
+				}
+			})
+		}
+	})
+}

